### ISSUE TITLE
HttpOperationException: Precondition Failed after quick messages

### ISSUE THREAD

**Issue by NanashiGIT** [Microsoft/BotBuilder#960]
Hello!
So i haven't noticed that exception before and im pretty sure i tried out that scenario:
When i send quick messages to my bot, i get the following exception (With a "Sorry. my bot has a problem")

> Microsoft.Rest.HttpOperationException: Precondition Failed    at Microsoft.Bot.Connector.ErrorHandling.HandleError<HttpOperationResponse`1 result|ObjectT>    at Microsoft.Bot.Connector.BotStateExtensions.<SetConversationDataAsync>d*9.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Internals.ConnectorStore.<Microsoft-Bot-Builder-Dialogs-Internals-IBotDataStore<Microsoft-Bot-Connector-BotData>-SaveAsync>d*3.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore.<Save>d*10.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore.<Microsoft-Bot-Builder-Dialogs-Internals-IBotDataStore<Microsoft-Bot-Connector-BotData>-FlushAsync>d*5.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Internals.BotDataBase1.<FlushAsync>d__9.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Internals.PersistentDialogTask.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__61.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Internals.PostUnhandledExceptionToUserTask.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d*5`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at Microsoft.Bot.Builder.Dialogs.Internals.PostUnhandledExceptionToUserTask.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d*5`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Conversation.<SendAsync>d*4.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at Microsoft.Bot.Builder.Dialogs.Conversation.<SendAsync>d*2.MoveNext() --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()    at Bot.MessagesController.<Post>d__5.MoveNext() in C
9:32 :\Workspace\MyBot\Bot\Controllers\MessagesController.cs:line 95

To cause that exception, i jsut have to send like 3 Messages within 1 or 2 seconds. The 2 last messages arrive at the bot while the first one is still being processed by the bot.
I'm pretty sure I've tried doing that before 3.1, but im not 100% sure.

Has there something changed how the Connector handles messages to a bot, while there is still a message being processed?
Is there a workaround for that?

By the way: The error doesnt occur with the emulator.

Greetings

**Comment by willportnoy** [comment]
Yes, we're using ETag consistency for writes to the bot data store.

https://github.com/Microsoft/BotBuilder/releases/tag/NuGet3.1.0
change is this commit: https://github.com/Microsoft/BotBuilder/commit/6ce11a957c0863676ebe6589f0be2c2c31d74de3

@msft-shahins for visibility

**Comment by NanashiGIT** [comment]
Uhm. So i'm not completely sure what an ETag is or what its good for.

So.... How to fix that exception? 
Is it a bug with the new etag approach or is there a way to fix it in my case?
Im saving 1 Object in the Bot data bag, if thats important.

Greetings

**Comment by willportnoy** [comment]
https://docs.botframework.com/en-us/technical-faq/#what-is-an-etag--how-does-it-relate-to-bot-data-bag-storage has some more information.

You can see why your bot may be taking a long time to answer, or you could choose to remove the etag optimistic concurrency control by using a different policy: https://github.com/Microsoft/BotBuilder/commit/6ce11a957c0863676ebe6589f0be2c2c31d74de3#diff-0275a9b9e0adcfef541649b85faac607R240

**Comment by NanashiGIT** [comment]
Okay. I dont really want to turn it off, because I think you had a reason to implement that policy.

But Im not sure how to handle that error. In the MessagesController, I start a new Luis-Dialog (my main dialog). When that Precondition-Error happens, i cant manage to catch that error inside the dialog. All I can do is catch that dialog at the call of it in the MessagesController.
But that results in the dialog being aborted/canceled and the user is not in it anymore.

I'm asuming my bot needs a bit longer, due to the connection with LUIS, but im pretty sure even the fastest Bot would throw that error, when I send multiple messages within a second. And that shouldnt result in the dialog being canceled.

Do you have any idea how to handle that properly?

By the way: I get the default Error message "Sorry, my bot code is having an issue." and then my catch in the messagesController is reached, so there is another try catch somewhere before i can manually handle the exception, which isnt a nice approach in my opinion.

**Comment by wiltodelta** [comment]
It helped me, but not sure of the correctness of that decision.

```
			var builder = new ContainerBuilder();

			builder.Register(c => new CachingBotDataStore(c.Resolve<ConnectorStore>(),
				CachingBotDataStoreConsistencyPolicy.LastWriteWins))
				.As<IBotDataStore<BotData>>()
				.AsSelf()
				.InstancePerLifetimeScope();

			builder.Update(Conversation.Container);
```

**Comment by NanashiGIT** [comment]
Hm where do you call that method?
In the messagesController when every message arrives or only once in some constructor?

**Comment by wiltodelta** [comment]
```
	public class Global : HttpApplication
	{
		void Application_Start()
		{
			BotConfig.UpdateConversationContainer();
		}
	}
```

```
	public static class BotConfig
	{
		public static void UpdateConversationContainer()
		{
			var builder = new ContainerBuilder();

			builder.Register(c => new CachingBotDataStore(c.Resolve<ConnectorStore>(),
				CachingBotDataStoreConsistencyPolicy.LastWriteWins))
				.As<IBotDataStore<BotData>>()
				.AsSelf()
				.InstancePerLifetimeScope();

			builder.Update(Conversation.Container);
		}
	}
```

**Comment by NanashiGIT** [comment]
Thanks. It works.

But I'm still interested in an answer to how to handle the mentioned exception. Because I assume that we will run into concurrency issues with that solution.

**Comment by willportnoy** [comment]
I'm working on some code that should help - I'll update this issue thread when it's ready.

**Comment by willportnoy** [comment]
This commit will help serialize message handling within the scope of a particular conversation and user: https://github.com/Microsoft/BotBuilder/commit/8f441469e1e80b5d7712cbf81bc3eacc9f71e3a4

**Comment by NanashiGIT** [comment]
Uhm.
So i dont really understand your code, but i guess its really helpful :)

So once its released, how will we be able to catch an exception or how will the new behaviour be like?

**Comment by willportnoy** [comment]
With this new code, you won't have to do anything different (if you run your bot on a single node) - it will serialize the execution of your bot's code (for a particular conversation/user) so that only one message will be handled at a time (for that particular conversation/user).

**Comment by willportnoy** [comment]
This has been released: https://github.com/Microsoft/BotBuilder/releases/tag/NuGet3.2.0

I'm going to close this issue, but feel free to open a new issue for further discussion.

**Comment by alokraj68** [comment]
Hi @willportnoy 
Got kind of the same issue.
Attaching my error log.

```
Microsoft.Rest.HttpOperationException: The remote server returned an error: (412) Precondition Failed.
   at Microsoft.Bot.Connector.ErrorHandling.<HandleErrorAsync>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Connector.BotStateExtensions.<SetUserDataAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.ConnectorStore.<Microsoft-Bot-Builder-Dialogs-Internals-IBotDataStore<Microsoft-Bot-Connector-BotData>-SaveAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore.<Save>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore.<Microsoft-Bot-Builder-Dialogs-Internals-IBotDataStore<Microsoft-Bot-Connector-BotData>-FlushAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.BotDataBase`1.<FlushAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.DialogTaskManagerBotDataLoader.<FlushAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.PersistentDialogTask.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.ExceptionTranslationDialogTask.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.SerializeByConversation.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.PostUnhandledExceptionToUser.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Bot.Builder.Dialogs.Internals.PostUnhandledExceptionToUser.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.LogPostToBot.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Conversation.<SendAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Conversation.<SendAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at *******.<Post>d__1.MoveNext() in ***********\MessagesController.cs:line 51
```

**Comment by willportnoy** [comment]
@alokraj68 these answers might be helpful:

https://docs.botframework.com/en-us/technical-faq/#what-causes-precondition-failed-412-or-conflict-409-http-errors

**Comment by alokraj68** [comment]
Already checked them, it's more like so much stuff that doesn't give an answer.

**Comment by willportnoy** [comment]
@alokraj68 the answer is to decide how you want to handle concurrent requests across multiple replicas.  for single replicas, the builder SDK will already serialize message handling for a given conversation.  for multiple replicas, you might consider using a redis lease per conversation.

**Comment by Unders0n** [comment]
Still have issues with “precondition failed” error. Case is simple: I have a form dialog and I want for user only once fill this form, then it might be ignored. So I cant find a better way that to set some flag or state in Userdata and check it in messageController:
```
var state = activity.GetStateClient();
                var userData = await state.BotState.GetUserDataAsync(activity.ChannelId, activity.From.Id);
                if (userData.GetProperty<string>("state") != "FilledWelcomePoll")
                {
                    await Conversation.SendAsync(activity, MakeRootDialog); //form dialog
                }
```

<form result callback>
…
```
var act = context.Activity;
            var state = act.GetStateClient();
            var userData = await state.BotState.GetUserDataAsync(act.ChannelId, act.From.Id);
            userData.SetProperty("state", "FilledWelcomePoll");
            await state.BotState.SetUserDataAsync(act.ChannelId, act.From.Id, userData);
            context.Done(answers);
```

So if I left everything’s on default I get "precondition failed" on await Conversation.SendAsync when form is filled. If I set CachingBotDataStoreConsistencyPolicy.LastWriteWins in container save just doesn’t seems to happen.

**Comment by JobaDiniz** [comment]
@wiltodelta your code throws an exception for me
```
builder.Register(c => new CachingBotDataStore(c.Resolve<ConnectorStore>(),
                CachingBotDataStoreConsistencyPolicy.LastWriteWins))
                .As<IBotDataStore<BotData>>()
                .AsSelf()
                .InstancePerLifetimeScope();
```

```
Autofac.Core.DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = IBotData (DelegateActivator), Services = [Microsoft.Bot.Builder.Dialogs.Internals.IBotData], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = JObjectBotData (ReflectionActivator), Services = [Microsoft.Bot.Builder.Dialogs.Internals.JObjectBotData (Microsoft.Bot.Builder.Dialogs.Internals.IBotData), Microsoft.Bot.Builder.Dialogs.Internals.JObjectBotData], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = CachingBotDataStore (DelegateActivator), Services = [Microsoft.Bot.Builder.Dialogs.Internals.IBotDataStore`1[[Microsoft.Bot.Connector.BotData, Microsoft.Bot.Connector, Version=3.11.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> The requested service 'Microsoft.Bot.Builder.Dialogs.Internals.ConnectorStore' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency. (See inner exception for details.) (See inner exception for details.) (See inner exception for details.) ---> Autofac.Core.DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = JObjectBotData (ReflectionActivator), Services = [Microsoft.Bot.Builder.Dialogs.Internals.JObjectBotData (Microsoft.Bot.Builder.Dialogs.Internals.IBotData), Microsoft.Bot.Builder.Dialogs.Internals.JObjectBotData], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = CachingBotDataStore (DelegateActivator), Services = [Microsoft.Bot.Builder.Dialogs.Internals.IBotDataStore`1[[Microsoft.Bot.Connector.BotData, Microsoft.Bot.Connector, Version=3.11.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> The requested service 'Microsoft.Bot.Builder.Dialogs.Internals.ConnectorStore' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency. (See inner exception for details.) (See inner exception for details.) ---> Autofac.Core.DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = CachingBotDataStore (DelegateActivator), Services = [Microsoft.Bot.Builder.Dialogs.Internals.IBotDataStore`1[[Microsoft.Bot.Connector.BotData, Microsoft.Bot.Connector, Version=3.11.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], Microsoft.Bot.Builder.Dialogs.Internals.CachingBotDataStore], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> The requested service 'Microsoft.Bot.Builder.Dialogs.Internals.ConnectorStore' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency. (See inner exception for details.) ---> Autofac.Core.Registration.ComponentNotRegisteredException: The requested service 'Microsoft.Bot.Builder.Dialogs.Internals.ConnectorStore' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.
   at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.Resolve[TService](IComponentContext context, IEnumerable`1 parameters)
   at WDG.Automation.Bot.Api.BotController.<.cctor>b__1(IComponentContext c) in e:\Source\joberto.diniz\Automation\WDG.Automation\Bot\WDG.Automation.Bot.Api\Controllers\BotController.cs:line 35
   at Autofac.Builder.RegistrationBuilder.<>c__DisplayClass0_0`1.<ForDelegate>b__0(IComponentContext c, IEnumerable`1 p)
   at Autofac.Core.Activators.Delegate.DelegateActivator.ActivateInstance(IComponentContext context, IEnumerable`1 parameters)
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   --- End of inner exception stack trace ---
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   at Autofac.Core.Lifetime.LifetimeScope.GetOrCreateAndShare(Guid id, Func`1 creator)
   at Autofac.Core.Resolving.InstanceLookup.Execute()
   at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, IComponentRegistration registration, IEnumerable`1 parameters)
   at Autofac.Core.Activators.Reflection.ConstructorParameterBinding.Instantiate()
   at Autofac.Core.Activators.Reflection.ReflectionActivator.ActivateInstance(IComponentContext context, IEnumerable`1 parameters)
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   --- End of inner exception stack trace ---
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   at Autofac.Core.Lifetime.LifetimeScope.GetOrCreateAndShare(Guid id, Func`1 creator)
   at Autofac.Core.Resolving.InstanceLookup.Execute()
   at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, IComponentRegistration registration, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.TryResolveService(IComponentContext context, Service service, IEnumerable`1 parameters, Object& instance)
   at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.ResolveKeyed[TService](IComponentContext context, Object serviceKey, IEnumerable`1 parameters)
   at Microsoft.Bot.Builder.Autofac.Base.Extensions.<>c__DisplayClass1_0`1.<RegisterAdapterChain>b__0(IComponentContext c)
   at Autofac.Builder.RegistrationBuilder.<>c__DisplayClass0_0`1.<ForDelegate>b__0(IComponentContext c, IEnumerable`1 p)
   at Autofac.Core.Activators.Delegate.DelegateActivator.ActivateInstance(IComponentContext context, IEnumerable`1 parameters)
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   --- End of inner exception stack trace ---
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   at Autofac.Core.Lifetime.LifetimeScope.GetOrCreateAndShare(Guid id, Func`1 creator)
   at Autofac.Core.Resolving.InstanceLookup.Execute()
   at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, IComponentRegistration registration, IEnumerable`1 parameters)
   at Autofac.Core.Resolving.ResolveOperation.Execute(IComponentRegistration registration, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.TryResolveService(IComponentContext context, Service service, IEnumerable`1 parameters, Object& instance)
   at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.Resolve[TService](IComponentContext context, IEnumerable`1 parameters)
   at WDG.Automation.Bot.Api.BotController.<Post>d__8.MoveNext() in e:\Source\joberto.diniz\Automation\WDG.Automation\Bot\WDG.Automation.Bot.Api\Controllers\BotController.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.TaskHelpersExtensions.<CastToObject>d__3`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()
```
